File: Problem_Solving/leetcode/linear_data_structure/Hash/347_Top_K_Frequent_Elements/topFrequent.cpp
/*
    347. Top K Frequent Elements
        Given an integer array nums and an integer k, 
        return the k most frequent elements. 
        You may return the answer in any order.

    Example 1:
        Input: nums = [1,1,1,2,2,3], k = 2
        Output: [1,2]
        Example 2:

        Input: nums = [1], k = 1
        Output: [1]

    Constraints:
        - 1 <= nums.length <= 105
        - -104 <= nums[i] <= 104
        - k is in the range [1, the number of unique elements in the array].
        - It is guaranteed that the answer is unique.

    Follow up: Your algorithm's time complexity must be better than O(n log n), where n is the array's size.
*/
#include <vector>
#include <unordered_map>
#include <queue>

// 테이블과 우선순위 큐를 사용해 상위 k번만큼 추출
std::vector<int> topKFrequent(std::vector<int> &nums, int k)
{
    std::unordered_map<int, int> map;
    // frequency
    for (int i : nums)
        map[i]++;

    std::vector<int> res;
    std::priority_queue<std::pair<int, int>> topk;
    for (auto itr : map)
        topk.push(std::make_pair(itr.second, itr.first));
    
    while (k--)
    {
        res.push_back(topk.top().second);
        topk.pop();
    }
    return res;
}

File: Problem_Solving/leetcode/linear_data_structure/Hash/347_Top_K_Frequent_Elements/topFrequent.py
# counter를 이용한 음수 순 추출
import collections
import heapq

def topKFrequent(self, nums: List[int], k: int) -> List[int]:
    freqs = collections.Counter(nums)
    freqs_heap = [] # priority queue
    # 힙에 음수로 삽입
    for f in freqs:
        # 빈도 수를 key로 해 삽입해준다.
        # 파이썬 heapq 모듈은 최소 힙(Min-Heap)만 지원하기에 음수로 저장
        heapq.heappush(freqs_heap, (-freqs[f], f))
        
    topk = list()
    # k번 만큼 추출, 최소 힙(Min Heap)이므로 가장 작은 음수 순으로 추출
    for _ in range(k):
        topk.append(heapq.heappop*(freqs_heap)[1])
        
    return topk